[FLINK-32751][collect] Fixes race condition between close and request handling processes in CollectSinkOperatorCoordinator

XComp opened the pull request:
:warning: I sliced the PR into reasonable commits. It might make sense to review the commits individually rather than the entire PR at once.

## What is the purpose of the change

The changes in https://github.com/apache/flink/pull/23180 caused more instabilities. Another investigation round revealed that we have, indeed, a race condition between the `CollectSinkOperatorCoordinator`'s close operation and the request handling. It can be that the `close()` call closes the socket while a request is handled. The request handler will only check whether a socket is null and reopen it if it's the case (see [CollectSinkOperatorCoordinator.java:160](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java#L160)).

This PR also improves the coverage of `CollectSinkOperatorCoordinatorTest`.

## Brief change log

* Introduces `running` volatile field in `CollectSinkOperatorCoordinator` and `connectionLock` to synchronize the race condition between `#close()` and `#handleRequestImpl(..)`
* Added an explicit cancellation of ongoing requests
* Introduces `SocketConnection` class to collect socket-related components in a single place.

## Verifying this change

* `CollectSinkOperatorCoordinatorTest.ServerThread` is replaced by `CollectSinkOperatorCoordinatorTest.TestingSocketServer`
  * Any obsolete test code (e.g. version forwarding and offset setting) was removed because related asserts only verified test code
  * `handle` methods were added to give explicit control over how the request should be handled
* Helper constructors were added to `CollectSinkOperatorCoordinator` (to be able to inject custom executors)
* New test cases were added:
  * `testNoAddress`: (Already existed before) tests the behavior if a no server is registered, yet.
  * `testCloseAfterRequestIsReceived`: Covers the actual bug and checks how the coordinator behaves if the coordinator is closed before handling the request
  * `testSuccessfulResponse`: Successful handling of the request by the server
  * `testServerSideClosingTheServerSocket`: the server shuts down entirely while processing a request
  * `testServerSideClosingTheAcceptingSocket`: checks how the coordinator behaves when the server side stops accepting requests before handling the request
  * `testServerSideNotResponding`: checks how the coordinator behaves if the server doesn't respond and the coordinator is closed
  * `testServerSideDisconnectWithReconnect`: checks the reconnect scenario (essentially a operator failover)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

zentol commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java:
```diff
@@ -152,9 +176,11 @@ private CoordinationResponse handleRequestImpl(
-            socketConnection = SocketConnection.create(socketTimeout, sinkAddress);
-            LOG.info("Sink connection established");
+        synchronized (connectionLock) {
+            if (running && socketConnection == null) {
```
otherwise we should exit eagerly instead of running into an NPE later on.

XComp replied:
Good point. This remark became obsolete because I removed the synchronization again.

zentol commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java:
```diff
@@ -152,9 +176,11 @@ private CoordinationResponse handleRequestImpl(
-        if (socketConnection == null) {
-            socketConnection = SocketConnection.create(socketTimeout, sinkAddress);
-            LOG.info("Sink connection established");
+        synchronized (connectionLock) {
```
Wondering if this shouldn't be a larger block.
Does the client handle it properly when the connection breaks down while it is reading data?

What's the desired behavior here?

XComp replied:
I didn't consider this one because of the [findings below](https://github.com/apache/flink/pull/23296#discussion_r1307474170)

zentol commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -173,72 +174,192 @@ private static void assertResponse(
+
+        private CompletableFuture<SocketConnection> connectSocket() {
+            if (!connectionFuture.isDone()) {
+                FutureUtils.forward(acceptSocketAsync(), connectionFuture);
```
Feels weird to set up this forwarding here instead of whenever the `connectionFuture` is reset (in the constructor and closeAcceptingSocket)

XComp replied:
That remark, I don't get. We want to have the same future that's accessible all the time (that's why `new CompletableFuture` here and in the constructor) but we do want to trigger the socket listening later on. How would you implement that differently. It feels natural to me. But I might have missed your point.

zentol replied:
I guess my point is why accepting the connection is a manual step in the first place.

This can also leak connections if connectSocket is called multiple times before the connectionFuture is being completed.

zentol commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
 
-            // server closes here
-            failedResponseFuture =
+                dataFuture.complete(Collections.singletonList(Row.of(123, "abc")));
```
why do we touch the data-future again? It's a server-side thing and should have no effect on the test results.

XComp replied:
fair enough. I removed these lines.

zentol commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
-            failedResponseFuture =
+                dataFuture.complete(Collections.singletonList(Row.of(123, "abc")));
+
+                FlinkAssertions.assertThatFuture(serverSideProcess)
+                        .eventuallyFailsWith(ExecutionException.class)
+                        .withCauseInstanceOf(SocketException.class);
```
This seems to primarily be testing the `TestingSocketServer`.

zentol replied:
I would completely drop the last 2 assertion blocks

XComp replied:
I also removed the lines above (see [comment](https://github.com/apache/flink/pull/23296#discussion_r1308284339)). Injecting the data is also not necessary in `testClosingTheListeningSocketInTheSinkFunction` (aka `testServerSideClosingTheAcceptingSocket`)

zentol commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
+                    .as("The connection to the client should be established.")
+                    .eventuallySucceeds();
+
+            // simulation of the server side failing while the request is handled on the server side
```
Maybe we should drop the client/server terminology because afaict we have 3 participants; the actual client (e.g., CLI), the coordinator and the sink operator (== TestingSocketServer), creating 2 client/server pairs.
It's just really confusing.

XComp replied:
As mentioend [above](https://github.com/apache/flink/pull/23296#discussion_r1308280794):
> I refactored the method and variable names to use coordinator and sinkFunction. Additionally, TestingSocketServer is renamed into TestingSinkFunction

XComp commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java:
```diff
@@ -83,7 +104,10 @@ public void start() throws Exception {
         LOG.info("Closing the CollectSinkOperatorCoordinator.");
+        running = false;
         this.executorService.shutdownNow();
+        ongoingRequests.forEach(ft -> ft.cancel(true));
```
Yeah, you're right. I don't know why but I remember that I assumed in the previous PR that shutdown would trigger a cancellation of the queued tasks. But that's not the case. The guarding of the `SocketConnection` and the `running` are obsolete: The running tasks are handled by `shutdownNow` and the queued tasks are cancelled before executing them. Therefore, no concurrency happens when resetting the `SocketConnection`.

That also explains the more frequent test failures: The previous `shutdown` call tried to handle the ongoing requests sequentially. The `shutdownNow` seem to have cancelled the ongoing task but queued tasks were executed afterwards which lead to the re-instantiation of the `SocketConnection`

XComp commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
+    }
+
+    @Test
+    void testServerSideNotResponding() throws Exception {
```
I refactored the method and variable names to use `coordinator` and `sinkFunction`. Additionally, `TestingSocketServer` is renamed into `TestingSinkFunction`

XComp commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinatorTest.java:
```diff
@@ -105,54 +102,147 @@ void testCloseAfterRequestIsReceived() throws Exception {
+                FlinkAssertions.assertThatFuture(socketServer.handleRequestWithoutResponse())
+                        .eventuallySucceeds();
+            }
+            assertEmptyResponseGeneratedFromServer(responseFuture);
```
I removed the try/catch blocks where I felt that having the explicit close call being part of the test code helps the readability.

XComp commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java:
```diff
@@ -83,7 +104,10 @@ public void start() throws Exception {
     public void close() throws Exception {
         LOG.info("Closing the CollectSinkOperatorCoordinator.");
+        running = false;
         this.executorService.shutdownNow();
```
I removed this due to the findings from [the discussion below](https://github.com/apache/flink/pull/23296#discussion_r1307474170)

XComp commented:
I addressed all comments and force-pushed the reorganized commits. Thanks for your review, @zentol. :-) Much appreciated. I couldn't reproduce any test instability locally with 200000 test runs. I will go ahead and create backports for the other branches.
